crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2418,7 +2416,23 @@ impl<'db> TypeInferenceBuilder<'db> {
                     None
                 };
 
-                bindings_ty(self.db, definitions, unbound_ty)
+                let ty = bindings_ty(self.db, definitions, unbound_ty);
+
+                if ty.is_unbound() {
+                    self.add_diagnostic(
+                        name.into(),
+                        "unresolved-reference",
+                        format_args!("Name `{id}` used when not defined"),
+                    );
+                } else if ty.may_be_unbound(self.db) {
+                    self.add_diagnostic(
+                        name.into(),
+                        "possibly-unresolved-reference",
```

Comment:
@AlexWaygood what's the motivation for having two different rule codes? I'm in favor of specific rules, but I'm not sure if we need those fine granular diagnostic codes.

Reply:
Partly the motivation I gave in https://github.com/astral-sh/ruff/pull/13922#issuecomment-2438372482: it's much easier to test them robustly if we have two different codes.

In the long term, I do also think it will be better for users if these are separate codes. `unresolved-reference` should be a code that we have very high confidence in; I would expect it to be seen as a massive code smell if somebody added a `# knot: ignore[unresolved-reference]` comment in their Python code. `possibly-unresolved-reference` will by its nature have a lot more false positives, so ignoring the error might sometimes be reasonable. (For example: a variable might be defined in the body of a loop. From red-knot's perspective, all loops could potentially have zero iterations if the loop loops over an empty sequence — so variables defined as `T` in a loop are always inferred as being `Unbound | T` by red-knot from the perspective of code following the loop. But sometimes the user knows for sure that a certain loop will always have >=1 iterations, meaning they can be certain that the variable is defined after the loop even if red-knot can't be. In that situation, it might be reasonable to suppress the `possibly-unresolved-reference` error that red-knot emits.)